src/librustc_mir/borrow_check/nll/explain_borrow/mod.rs

```diff
@@ -59,17 +60,33 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
 
                     Cause::DropVar(local, location) => {
                         match find_drop_use(mir, regioncx, borrow, location, local) {
-                            Some(p) => {
-                                let local_name = mir.local_decls[local].name.unwrap();
-
-                                err.span_label(
-                                    mir.source_info(p).span,
-                                    format!(
-                                        "borrow later used here, when `{}` is dropped",
-                                        local_name
-                                    ),
-                                );
-                            }
+                            Some(p) => match &mir.local_decls[local].name {
+                                Some(local_name) => {
+                                    err.span_label(
+                                        mir.source_info(p).span,
+                                        format!(
+                                            "borrow later used here, when `{}` is dropped",
+                                            local_name
+                                        ),
+                                    );
+                                }
+                                None => {
+                                    let locations = mir.find_assignments(local);
+
+                                    if locations.len() == 1 {
+                                        err.span_label(
+                                            mir.source_info(locations[0]).span,
+                                            "borrow may end up in a temporary, created here"
+                                        );
+                                    }
+
+                                    err.span_label(
+                                        mir.source_info(p).span,
```

Comment:
actually maybe we can just do `mir.local_decls[local].source_info.span` here and not bother to search for locals?